rules/dispatch.go

```diff
@@ -387,3 +333,193 @@ func (r *DefaultRuleRenderer) dispatchChains(
 
 	return chains
 }
+
+func (r *DefaultRuleRenderer) dispatchEndPointMarkChains(
+	wlNames []string,
+	hepNames []string,
+	epMarkMapper EndpointMarkMapper,
+	setMarkPfx,
+	wlFromMarkPfx,
+	hepFromMarkPfx,
+	dispatchSetMarkEndpointChainName,
+	dispatchFromMarkEndpointChainName string,
+) []*Chain {
+	names := append(wlNames, hepNames...)
+	// Sort interface names so that rules in the dispatch chain are ordered deterministically.
+	// Otherwise we would reprogram the dispatch chain when there is no real change.
+	sort.Strings(names)
+	log.WithField("ifaceNames", names).Debug("Rendering endpoint mark dispatch chains")
+
+	rootSetMarkRules := make([]Rule, 0)
+	rootFromMarkRules := make([]Rule, 0)
+
+	// start rendering set mark rules.
+	// Since there can be >100 endpoints, putting them in a single list adds some latency to
+	// endpoints that are later in the chain.  To reduce that impact, we build a shallow tree of
+	// chains based on the prefixes of the chains.
+	commonPrefix, prefixes, prefixToNames := r.divideEndpointNamesToPrefixTree(names)
+
+	// Now, iterate over the prefixes.  If there are multiple names in a prefix, we render a
+	// child chain for that prefix.  Otherwise, we render the rule directly to avoid the cost
+	// of an extra goto.
+	var chains []*Chain
+	for _, prefix := range prefixes {
```

Comment:
I find this much clearer than the old code but it might be worth a try to put his into a utility function to build a single chain: `buildDispatchChain(rootPrefix, childPrefix string, getActionForEndpoint func(name string) iptables.Action, prefixes <...etc>)`

Then above, you'd call it twice, to build the to/from chains, whereas here, you'd only call it once.

Reply:
We need to call it twice too for workload and host endpoint. It is a bit tricky to combine those but it is doable. I created a `buildSingleDispatchChains` and return `childChains, rootChain rootRule`. The caller can make tweeks (merge, add rules to rootRules) as needed.